pip/src/interpreter.rs

```diff
@@ -494,6 +574,89 @@ impl Interpreter {
     }
 }
 
+/// Check if the given item supports interop with Python by verifying the input and output types.
+fn item_supports_interop(interpreter: &interpret::Interpreter, id: StoreItemId) -> bool {
+    let (input_ty, output_ty) = interpreter
+        .get_callable_tys(id)
+        .expect("item should exist in package");
+    type_supports_interop(&input_ty) && type_supports_interop(&output_ty)
+}
+
+/// Confirm a Q# type supports with interop with Python, meaning our code knows how to convert it back and forth
+/// across the interop boundary.
+fn type_supports_interop(ty: &Ty) -> bool {
+    match ty {
+        Ty::Prim(prim_ty) => match prim_ty {
+            Prim::Pauli
+            | Prim::BigInt
+            | Prim::Bool
+            | Prim::Double
+            | Prim::Int
+            | Prim::String
+            | Prim::Result => true,
+            Prim::Qubit | Prim::Range | Prim::RangeTo | Prim::RangeFrom | Prim::RangeFull => false,
+        },
+        Ty::Tuple(tup) => tup.iter().all(type_supports_interop),
+        Ty::Array(ty) => type_supports_interop(ty),
+        _ => false,
+    }
+}
+
+/// Given a type, convert a Python object into a Q# value of that type. This will recur through tuples and arrays,
+/// and will return an error if the type is not supported or the object cannot be converted.
+fn convert_obj_with_ty(py: Python, obj: &PyObject, ty: &Ty) -> PyResult<Value> {
+    match ty {
+        Ty::Prim(prim_ty) => match prim_ty {
+            Prim::BigInt => Ok(Value::BigInt(obj.extract::<BigInt>(py)?)),
+            Prim::Bool => Ok(Value::Bool(obj.extract::<bool>(py)?)),
+            Prim::Double => Ok(Value::Double(obj.extract::<f64>(py)?)),
+            Prim::Int => Ok(Value::Int(obj.extract::<i64>(py)?)),
+            Prim::String => Ok(Value::String(obj.extract::<String>(py)?.into())),
+            Prim::Result => Ok(Value::Result(qsc::interpret::Result::Val(
+                obj.extract::<Result>(py)? == Result::One,
+            ))),
+            Prim::Pauli => Ok(Value::Pauli(match obj.extract::<Pauli>(py)? {
+                Pauli::I => fir::Pauli::I,
+                Pauli::X => fir::Pauli::X,
+                Pauli::Y => fir::Pauli::Y,
+                Pauli::Z => fir::Pauli::Z,
+            })),
+            Prim::Qubit | Prim::Range | Prim::RangeTo | Prim::RangeFrom | Prim::RangeFull => Err(
+                PyException::new_err(format!("unhandled primitive input type: {prim_ty:?}")),
+            ),
```

Comment:
Can we just avoid projecting Q# callables that take unmarshallable types as input? (Correct me if we're already doing that - I see that we're doing it for return types, but I don't see it being done for input types).

And once we do that, this should be a panic.

Reply:
The change already includes checking of both input and output (see the `item_supports_interop` function, which checks both types). These were nice errors for when the other code would try converting anything it found, but you are right in that these paths should never get it now. So when it's just for future us in developing newer features a panic is good enough, and panics are actually reported really nicely in Python via pyO3 so that makes investigation easy. I'll switch these!